### PR TITLE
Wip mysql cdc fix

### PR DESCRIPTION
**Pull request by dartartem**

### CODE REVIEW

File: eventuate-local-java-cdc-connector-mysql-binlog/src/main/java/io/eventuate/local/mysql/binlog/MySqlBinaryLogClient.java [hunk 1]
```diff
@@ -126,71 +140,91 @@ protected void leaderStart() {
...
+    }
+  }
 
-          if (shouldHandleTable) {
-            tableMapEventByTableId.put(tableMapEvent.getTableId(), tableMapEvent);
-          } else {
-            tableMapEventByTableId.remove(tableMapEvent.getTableId());
-          }
+  private void handleRestart(Exception e) {
+    logger.error(e.getMessage(), e.getMessage());
```
Comment (cer) [inline eventuate-local-java-cdc-connector-mysql-binlog/src/main/java/io/eventuate/local/mysql/binlog/MySqlBinaryLogClient.java]: Two things:

1. General problem: Don't repeat the message from the exception `logger.error(e.getMessage(), ... )`. Instead, logger.error("A useful error message like 'restarting due to exception')
2. Specific: the second argument to error() should be the exception.

File: eventuate-local-java-cdc-connector-mysql-binlog/src/main/java/io/eventuate/local/mysql/binlog/MySqlBinaryLogClient.java [hunk 2]
```diff
@@ -126,71 +140,91 @@ protected void leaderStart() {
...
+    } catch (InterruptedException e) {
+      handleProcessingFailException(e);
+    }
+  }
 
-          boolean shouldHandleTable = binlogEntryHandlers
-                  .stream()
-                  .map(BinlogEntryHandler::getSchemaAndTable)
-                  .anyMatch(schemaAndTable::equals);
+  private void listenBinlogEventsHandleErrors(Event event, Optional<BinlogFileOffset> binlogFileOffset) {
```
Comment (cer) [inline eventuate-local-java-cdc-connector-mysql-binlog/src/main/java/io/eventuate/local/mysql/binlog/MySqlBinaryLogClient.java]: This method is  badly named. `handleBinlogEventWithErrorHandling` would be better.

File: eventuate-local-java-cdc-connector-mysql-binlog/src/main/java/io/eventuate/local/mysql/binlog/MySqlBinaryLogClient.java [hunk 3]
```diff
@@ -126,71 +140,91 @@ protected void leaderStart() {
...
+    publishingException = Optional.of(e);
+    leaderSelector.stop();
+    callbackOnStop = Optional.of(() -> {
+      callbackOnStop = Optional.empty();
+      start();
+    });
+  }
 
-          dbLogMetrics.onBinlogEntryProcessed();
+  private void listenBinlogEvents(Event event, Optional<BinlogFileOffset> binlogFileOffset) {
```
Comment (cer) [inline eventuate-local-java-cdc-connector-mysql-binlog/src/main/java/io/eventuate/local/mysql/binlog/MySqlBinaryLogClient.java]: This method is  badly named. `handleBinlogEvent` would be better.

File: eventuate-local-java-cdc-connector-mysql-binlog/src/main/java/io/eventuate/local/mysql/binlog/MySqlBinaryLogClient.java (hunk 2 above)
Comment (cer) [inline eventuate-local-java-cdc-connector-mysql-binlog/src/main/java/io/eventuate/local/mysql/binlog/MySqlBinaryLogClient.java]: This method should return immediately if `publishingException.isPresent() == true`. i think it's possible that the binlog client might call listenBinlogEventsHandleErrors() again.

File: eventuate-local-java-cdc-connector-mysql-binlog/src/main/java/io/eventuate/local/mysql/binlog/MySqlBinaryLogClient.java [hunk 4]
```diff
@@ -240,6 +274,7 @@ private void handleWriteRowsEvent(Event event, Optional<BinlogFileOffset> starti
 
     onEventReceived();
     saveOffset(event);
+    publishingException = Optional.empty();
```
Comment (cer) [inline eventuate-local-java-cdc-connector-mysql-binlog/src/main/java/io/eventuate/local/mysql/binlog/MySqlBinaryLogClient.java]: It think `publishingException = Optional.empty();` should only be done in leadershipStart()

File: eventuate-local-java-cdc-connector-mysql-binlog/src/main/java/io/eventuate/local/mysql/binlog/MySqlBinaryLogClient.java (hunk 1 above)
Comment (dartartem) [reply eventuate-local-java-cdc-connector-mysql-binlog/src/main/java/io/eventuate/local/mysql/binlog/MySqlBinaryLogClient.java]: Ok, I will fix

File: eventuate-local-java-cdc-connector-mysql-binlog/src/main/java/io/eventuate/local/mysql/binlog/MySqlBinaryLogClient.java (hunk 2 above)
Comment (dartartem) [reply eventuate-local-java-cdc-connector-mysql-binlog/src/main/java/io/eventuate/local/mysql/binlog/MySqlBinaryLogClient.java]: ok

File: eventuate-local-java-cdc-connector-mysql-binlog/src/main/java/io/eventuate/local/mysql/binlog/MySqlBinaryLogClient.java (hunk 3 above)
Comment (dartartem) [reply eventuate-local-java-cdc-connector-mysql-binlog/src/main/java/io/eventuate/local/mysql/binlog/MySqlBinaryLogClient.java]: ok

File: eventuate-local-java-cdc-connector-mysql-binlog/src/main/java/io/eventuate/local/mysql/binlog/MySqlBinaryLogClient.java (hunk 2 above)
Comment (dartartem) [reply eventuate-local-java-cdc-connector-mysql-binlog/src/main/java/io/eventuate/local/mysql/binlog/MySqlBinaryLogClient.java]: I will fix, but as I know client just stops listen events when exception is thrown

File: eventuate-local-java-cdc-connector-mysql-binlog/src/main/java/io/eventuate/local/mysql/binlog/MySqlBinaryLogClient.java (hunk 4 above)
Comment (dartartem) [reply eventuate-local-java-cdc-connector-mysql-binlog/src/main/java/io/eventuate/local/mysql/binlog/MySqlBinaryLogClient.java]: ok

File: eventuate-local-java-cdc-connector-common/src/main/java/io/eventuate/local/common/BinlogEntryReaderLeadership.java [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    this.binlogEntryReader = binlogEntryReader;
+
+    binlogEntryReader.setRestartCallback(this::restart);
+  }
+
+  public void start() {
+    eventuateLeaderSelector = leaderSelectorFactory.create(leaderLockId,
+            UUID.randomUUID().toString(),
+            () -> {
+              leader = false;
```
[the remark this answers is not in the document]
Comment (dartartem) [reply eventuate-local-java-cdc-connector-common/src/main/java/io/eventuate/local/common/BinlogEntryReaderLeadership.java]: Chris, sorry, but I do not understand the question, could you please clarify what do you mean?

File: eventuate-local-java-cdc-connector-common/src/main/java/io/eventuate/local/common/BinlogEntryReaderLeadership.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+  public BinlogEntryReader getBinlogEntryReader() {
+    return binlogEntryReader;
+  }
+
+  public boolean isLeader() {
+    return leader;
+  }
+
+  private void restart() {
```
[the remark this answers is not in the document]
Comment (dartartem) [reply eventuate-local-java-cdc-connector-common/src/main/java/io/eventuate/local/common/BinlogEntryReaderLeadership.java]: Ok, I will work on fix

File: eventuate-local-java-cdc-connector-common/src/main/java/io/eventuate/local/common/BinlogEntryReaderLeadership.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+              actionOnStop.ifPresent(Runnable::run);
+            });
+
+    eventuateLeaderSelector.start();
+  }
+
+  public void stop() {
+    eventuateLeaderSelector.stop();
+    binlogEntryReader.stop();
+    binlogEntryReader.clearBinlogEntryHandlers();
```
[the remark this answers is not in the document]
Comment (dartartem) [reply eventuate-local-java-cdc-connector-common/src/main/java/io/eventuate/local/common/BinlogEntryReaderLeadership.java]: ok

[the remark this answers is not in the document]
Comment (dartartem) [reply eventuate-local-java-cdc-connector-common/src/main/java/io/eventuate/local/common/BinlogEntryReaderLeadership.java]: ok

[the remark this answers is not in the document]
Comment (dartartem) [reply eventuate-local-java-cdc-connector-common/src/main/java/io/eventuate/local/common/BinlogEntryReaderLeadership.java]: The idea was that reader requests restart, and in that case handlers are not removed.

File: eventuate-local-java-cdc-connector-unified/src/main/java/io/eventuate/local/unified/cdc/pipeline/common/BinlogEntryReaderLeadershipProvider.java [hunk 8]
```diff
@@ -0,0 +1,32 @@
+package io.eventuate.local.unified.cdc.pipeline.common;
+
+import io.eventuate.local.common.BinlogEntryReaderLeadership;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+
+public class BinlogEntryReaderLeadershipProvider {
```
[the remark this answers is not in the document]
Comment (dartartem) [reply eventuate-local-java-cdc-connector-unified/src/main/java/io/eventuate/local/unified/cdc/pipeline/common/BinlogEntryReaderLeadershipProvider.java]: yes

[the remark this answers is not in the document]
Comment (dartartem) [reply eventuate-local-java-cdc-connector-unified/src/main/java/io/eventuate/local/unified/cdc/pipeline/common/BinlogEntryReaderLeadershipProvider.java]: Do you think that it should be somehow improved?

[the remark this answers is not in the document]
Comment (dartartem) [reply eventuate-local-java-cdc-connector-unified/src/main/java/io/eventuate/local/unified/cdc/pipeline/common/BinlogEntryReaderLeadershipProvider.java]: ok, I will rename it back

File: eventuate-local-java-cdc-connector-unified/src/main/java/io/eventuate/local/unified/cdc/pipeline/common/BinlogEntryReaderLeadershipProvider.java [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+
+  private Map<String, BinlogEntryReaderLeadership> clients = new HashMap<>();
+
+  public void addBinlogEntryReaderLeadership(String name, BinlogEntryReaderLeadership binlogEntryReaderLeadership) {
```
[the remark this answers is not in the document]
Comment (dartartem) [reply eventuate-local-java-cdc-connector-unified/src/main/java/io/eventuate/local/unified/cdc/pipeline/common/BinlogEntryReaderLeadershipProvider.java]: ok

**Pull request merged**